File: example-codes/airline-management-system/flight_schedule.py
class WeeklySchedule:
    def __init__(self, day_of_week, departure_time):
        self.__day_of_week = day_of_week
        self.__departure_time = departure_time


class CustomSchedule:
    def __init__(self, custom_date, departure_time):
        self.__custom_date = custom_date
        self.__departure_time = departure_time


class Flight:
    def __init__(self, flight_number, departure, arrival, duration_in_minutes):
        self.__flight_number = flight_number
        self.__departure = departure
        self.__arrival = arrival
        self.__duration_in_minutes = duration_in_minutes

        self.__weekly_schedules = []
        self.__custom_schedules = []
        self.__flight_instances = []


class FlightInstance:
    def __init__(self, departure_time, gate, status, aircraft):
        self.__departure_time = departure_time
        self.__gate = gate
        self.__status = status
        self.__aircraft = aircraft

    def cancel(self):
        None

    def update_status(self, status):
        None


class FlightReservation:
    def __init__(self, reservation_number, flight, aircraft, creation_date, status):
        self.__reservation_number = reservation_number
        self.__flight = flight
        self.__seat_map = {}
        self.__creation_date = creation_date
        self.__status = status

    def fetch_reservation_details(self, reservation_number):
        None

    def get_passengers(self):
        None


class Itinerary:
    def __init__(self, customer_id, starting_airport, final_airport, creation_date):
        self.__customer_id = customer_id
        self.__starting_airport = starting_airport
        self.__final_airport = final_airport
        self.__creation_date = creation_date
        self.__reservations = []

    def get_reservations(self):
        None

    def make_reservation(self):
        None

    def make_payment(self):
        None

File: example-codes/airline-management-system/java/FlightSchedule.java
package com.airline;

import java.util.*;

/**
 * Handles scheduling of flights (weekly, custom), reservations, and itineraries.
 * This is the core logic of the system.
 */
public class FlightSchedule {

    // Represents recurring weekly flight schedules
    public static class WeeklySchedule {
        private String dayOfWeek;
        private String departureTime;
        public WeeklySchedule(String dayOfWeek, String departureTime) {
            this.dayOfWeek = dayOfWeek;
            this.departureTime = departureTime;
        }
    }

    // Represents special one-time flight schedules
    public static class CustomSchedule {
        private Date customDate;
        private String departureTime;
        public CustomSchedule(Date customDate, String departureTime) {
            this.customDate = customDate;
            this.departureTime = departureTime;
        }
    }

    // Represents a flight definition (without specific dates)
    public static class Flight {
        private String flightNumber;
        private Airport departure;
        private Airport arrival;
        private int durationInMinutes;
        private List<WeeklySchedule> weeklySchedules = new ArrayList<>();
        private List<CustomSchedule> customSchedules = new ArrayList<>();
        private List<FlightInstance> flightInstances = new ArrayList<>();

        public Flight(String flightNumber, Airport departure, Airport arrival, int durationInMinutes) {
            this.flightNumber = flightNumber;
            this.departure = departure;
            this.arrival = arrival;
            this.durationInMinutes = durationInMinutes;
        }
    }

    // Represents a specific occurrence of a flight (e.g., on 12th Nov)
    public static class FlightInstance {
        private String departureTime;
        private String gate;
        private Constants.FlightStatus status;
        private Aircraft aircraft;

        public FlightInstance(String departureTime, String gate, Constants.FlightStatus status, Aircraft aircraft) {
            this.departureTime = departureTime;
            this.gate = gate;
            this.status = status;
            this.aircraft = aircraft;
        }

        public void cancel() {
            this.status = Constants.FlightStatus.CANCELLED;
        }

        public void updateStatus(Constants.FlightStatus status) {
            this.status = status;
        }
    }

    // Represents a reservation made by a customer for a specific flight
    public static class FlightReservation {
        private String reservationNumber;
        private Flight flight;
        private Aircraft aircraft;
        private Date creationDate;
        private Constants.ReservationStatus status;

        public FlightReservation(String reservationNumber, Flight flight, Aircraft aircraft, Date creationDate, Constants.ReservationStatus status) {
            this.reservationNumber = reservationNumber;
            this.flight = flight;
            this.aircraft = aircraft;
            this.creationDate = creationDate;
            this.status = status;
        }
    }

    // Represents a customer's complete travel itinerary (can include multiple reservations)
    public static class Itinerary {
        private String customerId;
        private Airport startingAirport;
        private Airport finalAirport;
        private Date creationDate;
        private List<FlightReservation> reservations = new ArrayList<>();

        public Itinerary(String customerId, Airport startingAirport, Airport finalAirport, Date creationDate) {
            this.customerId = customerId;
            this.startingAirport = startingAirport;
            this.finalAirport = finalAirport;
            this.creationDate = creationDate;
        }
    }
}
